[jereldlimjy] iP

jereldlimjy opened the pull request:
# A heading

- hello world
- bye world

1. Hello!
2. Bye!

Use `git status` to list all new or modified files that haven't been committed.

```
git status
git add
git commit
```

This site was built using [GitHub Pages](https://pages.github.com/).

👍 

- [ ] eat lunch
- [ ] eat dinner

this is a **bold text**

this is a ~~strikethrough~~

> this is a quote

Commit:
Add Gradle support

jereldlimjy committed:
add greet, echo, exit

jereldlimjy committed:
add, list

jereldlimjy committed:
mark as done, task class

jereldlimjy committed:
todos, events and deadlines

jereldlimjy committed:
amended input and expected text files

jereldlimjy committed:
handle errors

jereldlimjy committed:
delete

jereldlimjy committed:
add save and load task list

jereldlimjy committed:
dates and times

jereldlimjy committed:
Merge branch 'branch-Level-7'

jereldlimjy committed:
fix merge conflict for branch-Level-8

laihuiqi commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,163 @@
+        // Greeting
+        printHorizontalLine();
+        System.out.println(
+            "Karen:\n" +
+            "Can I speak to your manager?\n" +
+            "Just kidding, this is Karen. How can I help you today?"
```
I like the personality of your chatbot! :smile:

laihuiqi commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,163 @@
+                    String type = command.split(" ")[0];
+                    String[] commandArr = command.split(" ", 2);
+
+                    switch (type) {
+                        case "todo":
```
Would you like to change the switch statement format like this to suit the coding standard need?
```suggestion
                    switch (type) {
                    case "todo":
```

laihuiqi commented on src/main/java/Save.java:
```diff
@@ -0,0 +1,74 @@
+    public static void saveTaskList(List<Task> taskList) throws IOException {
+        // check if directory exists
+        File dir = new File("./data");
+        if (!dir.exists()) dir.mkdirs();
```
Would you like to wrap this single if-statement with curly brackets?
```suggestion
        if (!dir.exists()) {
            dir.mkdirs();
        }
```

laihuiqi commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,163 @@
+                    Save.saveTaskList(tasks);
+
+                    System.out.println(
+                        "Okay okay, this has been removed:\n" +
+                        taskToDelete.toString() +
+                        "\nNow you have " + tasks.size() + " tasks left."
```
Maybe you can consider to put the '+' operator for every line breaks in a new line? :thinking:
```suggestion
                        "Okay okay, this has been removed:\n" 
                        + taskToDelete.toString() 
                        + "\nNow you have " + tasks.size() + " tasks left."
```

laihuiqi commented on text-ui-test/EXPECTED.TXT:
```diff
@@ -1,7 +1,44 @@
+[T][X] borrow book
+????????????????????????????????????????????????????????????????????????????????
+????????????????????????????????????????????????????????????????????????????????
+Karen:
+1. [T][X] borrow book
+2. [D][ ] return book (by: Sunday)
+3. [E][ ] project meeting (from: Mon 2pm to: 4pm)
```
I like your test case, it tests all the functions sufficiently! :tada:

laihuiqi reviewed (commented):
Nice to meet you! :wave:
I like the personality you gave to your chatbot! It is very interesting!! :+1::thumbsup:
However, I also noted a few minor coding standard violations and some class / method is too long. So, maybe you can consider splitting it to make them more readable?

jereldlimjy committed:
refactor code for more OOP

jereldlimjy committed:
add duke package

jereldlimjy committed:
Merge remote-tracking branch 'origin/add-gradle-support'

jereldlimjy committed:
set up gradle

jereldlimjy committed:
add junit tests

jereldlimjy committed:
add JavaDoc comments

jereldlimjy committed:
update code to comply with coding standard

jereldlimjy committed:
add find command

jereldlimjy committed:
fix setDone merge conflict

jereldlimjy committed:
Merge branch 'branch-Level-9'

astraxq commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,38 @@
+     */
+    public Deadline(String description, String by) throws DateTimeParseException {
+        super(description);
+        System.out.println(by);
```
Is there a need to print out `by`? Even if there is a need to print it out as a bot reply, it should not be done in the constructor side.

astraxq commented on src/main/java/duke/DeadlineCommand.java:
```diff
@@ -0,0 +1,27 @@
+     */
+    public DeadlineCommand(String description, String by) {
+        this.description = description;
+        this.by = by;
```
Create a `Deadline` Object field and run the constructor in the `DeadlineCommand` constructor as both `description` and `by` are not used anywhere else other than creating a `Deadline `object.

astraxq commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,73 @@
+     */
+    public void run() {
+        Scanner sc = new Scanner(System.in);
+        List<Task> taskList = tasks.getTasks();
```
`taskList` unused variable? 😨

astraxq commented on src/main/java/duke/EventCommand.java:
```diff
@@ -0,0 +1,30 @@
+    public EventCommand(String description, String from, String to) {
+        this.description = description;
+        this.from = from;
+        this.to = to;
```
Similar case as `DeadlineCommand`, just store an `Event` object instead of storing 3 `Strings` variable, keeps the code neat too.

astraxq commented on src/main/java/duke/FindCommand.java:
```diff
@@ -0,0 +1,20 @@
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        List<Task> taskList = tasks.getTasks();
+
+        ui.showFoundTasks(taskList.stream().filter(task -> task.toString().contains(keyword))
```
Perhaps implementing a `filter` method in `TaskList` class might be better. What if there's another command that requires some sort of filtering for the `taskList`?  

That would require more code to be written.

astraxq commented on src/main/java/duke/MarkCommand.java:
```diff
@@ -0,0 +1,27 @@
+        List<Task> taskList = tasks.getTasks();
+        Task currTask = taskList.get(taskNumber - 1);
+
+        currTask.setDone(true);
```
There's no need to create a new List instance, I personally feel that these 3 lines of code can be rewritten as:
```
List<Task> taskList = tasks.getTasks();
        Task currTask = taskList.get(taskNumber - 1);

        currTask.setDone(true);
```
```
tasks.get(taskNumber).setDone(true)
```
By implementing a  `get` method in `TaskList`, you can get a reference of the `Task` and call task methods to mark/unmark them.

astraxq commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,73 @@
+     * @throws DukeException If command cannot be parsed into a known command.
+     */
+    public static Command parse(String command) throws DukeException {
+        if (command.equals("list")) {
```
You can use `String.split(" ")` at the start, and use `command[0]` to handle different types of command. 

Subsequently, you can use the respective methods to handle the validity of the command.

astraxq commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,73 @@
+
+                String[] fromToArr = eventArr[1].split(" /to ");
+
+                if (fromToArr.length == 1) {
```
I believe validation can be done not inside the parse method, remember parse method main's job is to differentiate what type of commands it is and call the respective methods accordingly. 

I feel that there are too many things that parse is handling. Each method should be given a clear purpose.

astraxq commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,104 @@
+
+        if (f.exists()) {
+            Scanner s = new Scanner(f);
+            Task task;
```
Variables should be initialized where they are declared and they should be declared in the smallest scope possible.

astraxq commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,104 @@
+
+                if (taskSplit[0].equals("T")) {
+                    task = new Todo(taskSplit[2]);
+                } else if (taskSplit[0].equals("D")) {
```
You can try implementing a guard clause for this `if else` cases, as there will never be a case where there will be 2 triggered cases.

It makes your code look more neater too! 
Here's a [link](https://deviq.com/design-patterns/guard-clause) for reference!

E.g 
```
if (a) {
    doA();
}

if (b) {
    doB();
}

if (c) {
    doC();
}
```

astraxq commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,104 @@
+                        task.getDescription(),
+                        ((Event) task).getFrom(),
+                        ((Event) task).getTo());
+            }
```
Very nice way of saving it into the .txt file!

astraxq commented on src/main/java/duke/TodoCommand.java:
```diff
@@ -0,0 +1,24 @@
+     * @param description Description of the todo task.
+     */
+    public TodoCommand(String description) {
+        this.description = description;
```
Similarly to the other Task commands, just create the object in the constructor itself.

astraxq commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,34 @@
+     */
+    public void add(Task task) {
+        tasks.add(task);
+    }
```
Perhaps you can implement your delete/mark/unmark methods here too?

astraxq reviewed (commented):
Overall great job on the progress so far. Just some changes that I feel might make your code more readable/neater!

jereldlimjy committed:
remove unused variable

jereldlimjy committed:
add gui

jereldlimjy committed:
Add assertions

Add assertions to check for invariants.

This helps to minimise errors in the future.

jereldlimjy committed:
Improve code quality

Refactor code to improve quality based on feedback
from peers.

jereldlimjy committed:
Merge pull request #1 from jereldlimjy/branch-A-Assertions

Add assertions

jereldlimjy committed:
Merge branch 'master' into branch-A-CodeQuality

jereldlimjy committed:
Merge pull request #2 from jereldlimjy/branch-A-CodeQuality

Branch a code quality

jereldlimjy committed:
update duke and user images

jereldlimjy committed:
add command to mass delete tasks

jereldlimjy committed:
add background colours

jereldlimjy committed:
update user guide and add padding to dialog box

jereldlimjy committed:
update readme

jereldlimjy committed:
update header comment for Task

jereldlimjy committed:
add stats command

jereldlimjy committed:
update readme

jereldlimjy committed:
update readme
